subsys/usb/class/usb_dfu.c

```diff
@@ -480,7 +480,8 @@ static int dfu_class_handle_req(struct usb_setup_packet *pSetup,
 
 			if (len) {
 				ret = flash_read(dfu_data.flash_dev,
-						 dfu_data.flash_addr +
+						 dfu_data.flash_addr -
```

Comment:
> I was just confused why you want to do it, and I looked at the flash drivers for stm32, gecko and sam. Whatever this flash drivers do, I do not agree with the change here.

For what I understand, flash drivers take an offset relative to the beginning of the flash and not an absolute address. This is the case of at least the STM32 driver and the Kinetis MCU driver:
https://github.com/zephyrproject-rtos/zephyr/blob/7a85c19f1a01c56a0f0b2fed47b464471e348d6d/drivers/flash/soc_flash_mcux.c#L70

The NRF driver doesn't add any offset, but on the other hand the flash is mapped at address 0.

When writing data to the flash through `flash_img_buffered_write()` which in turns call `flash_write()` an offset relative to the beginning of the flash (`FLASH_AREA_IMAGE_1_OFFSET`) is used. On the other hand when the DFU code calls `flash_read`, it passes an absolute address as `dfu_data.flash_addr` is assigned either as `CONFIG_FLASH_BASE_ADDRESS +  FLASH_AREA_IMAGE_0_OFFSET`or  `CONFIG_FLASH_BASE_ADDRESS +  FLASH_AREA_IMAGE_1_OFFSET`.This is not consistent.

> It will at least break DFU on Kinetis MCUs (and NRF?, arduino101?).

Unfortunately I don't have a Kinetis board to try. Can you please tell me what would break on Kinetis MCUs, especially given CONFIG_FLASH_BASE_ADDRESS is set to 0 on those boards?

Reply:
@jfischer-phytec-iot any hint why it will break DFU on Kinetis MCUs?

Reply:
@aurel32 That's not as dramatic as I thought, I did not remember where FLASH_BASE_ADDRESS came from. As I see all the drivers where the reg for the flash node is zero just left it out, it not consistent. I am still confuse why it is not already DT_FLASH_BASE_ADDRESS, @galak ?

Reply:
> We would need the help from @galak to understand the DT part.

What's the question?  Is this about what the flash drivers are doing or something else?  We should be using CONFIG_FLASH_BASE_ADDRESS (that's what we generate from DTS).

Reply:
@galak Thanks for the confirmation, that's what is used in the patch.

Reply:
please consider discussion in  #12813